modelforge/potential/models.py

```diff
@@ -551,124 +546,114 @@ def apply_bwd(res, grads):
+        # ---------------------------
+        # general input manipulation
+        positions = data.positions
+        atomic_subsystem_indices = data.atomic_subsystem_indices
 
-        from modelforge.potential import _Implemented_NNPs
-        from modelforge.train.training import ModelTrainer
-
-        log.debug(f"{training_parameter=}")
-        log.debug(f"{potential_parameter=}")
-        log.debug(f"{dataset_parameter=}")
+        n = atomic_subsystem_indices.size(0)
 
-        if simulation_environment is None:
-            if runtime_parameter is None:
-                log.warning(
-                    "No runtime paramters or simulation_environment specified, defaulting to PyTorch"
-                )
+        # avoid reinitializing indices if they are already set and haven't changed
+        if self.indices.shape[0] != n:
+            # Generate a range of indices from 0 to n-1
+            self.indices = torch.arange(n, device=atomic_subsystem_indices.device)
 
-                simulation_environment = "PyTorch"
-            else:
-                simulation_environment = runtime_parameter.simulation_environment
-        # obtain model for training
-        if use == "training":
-            if simulation_environment == "JAX":
-                log.warning(
-                    "Training in JAX is not available. Falling back to PyTorch."
-                )
-            model = ModelTrainer(
-                potential_parameter=potential_parameter,
-                training_parameter=training_parameter,
-                dataset_parameter=dataset_parameter,
-                runtime_parameter=runtime_parameter,
-                potential_seed=potential_seed,
-            )
-            return model
-        # obtain model for inference
-        elif use == "inference":
-            model_type = potential_parameter.potential_name
-            nnp_class: Type = _Implemented_NNPs.get_neural_network_class(model_type)
-            model = nnp_class(
-                **potential_parameter.core_parameter.model_dump(),
-                postprocessing_parameter=potential_parameter.postprocessing_parameter.model_dump(),
-                dataset_statistic=dataset_statistic,
-                potential_seed=potential_seed,
+            # Create a meshgrid of indices
+            self.i_final_pairs, self.j_final_pairs = torch.meshgrid(
+                self.indices, self.indices, indexing="ij"
             )
-            if simulation_environment == "JAX":
-                return PyTorch2JAXConverter().convert_to_jax_model(model)
-            else:
-                return model
-        else:
-            raise NotImplementedError(f"Unsupported 'use' value: {use}")
+
+            # Filter out the diagonal elements (where i == j)
+            mask = self.i_final_pairs != self.j_final_pairs
+            self.i_final_pairs = self.i_final_pairs[mask]
+            self.j_final_pairs = self.j_final_pairs[mask]
+        # calculate r_ij and d_ij
```

Comment:
I'll make modifications in a separate PR after this is merged to speed things up (e.g., only using unique pairs and copying to get non-unique).  I'll note this takes in only_unique_pairs as an option, but doesn't do anything with it (but should).

Reply:
I noticed that there is no effect if I set `only_unique_pairs` to `True`.

Reply:
I can push a mini fix for this and a test. I fixed it in the nlist branch I started to feed into this. But rather fix this quickly in here otherwise ani will not work correctly in inference. Which would not be good.

Reply:
thanks!